torchmdnet/neighbors/neighbors_cuda_cell.cuh

```diff
@@ -0,0 +1,409 @@
+/* Raul P. Pelaez 2023. Batched cell list neighbor list implementation for CUDA.
+
+ */
+#ifndef NEIGHBOR_CUDA_CELL_H
+#define NEIGHBOR_CUDA_CELL_H
+#include "common.cuh"
+#include <thrust/extrema.h>
+
+/*
+ * @brief Calculates the cell dimensions for a given box size and cutoff
+ * @param box_size The box size
+ * @param cutoff The cutoff
+ * @return The cell dimensions
+ */
+template <typename scalar_t>
+__host__ __device__ int3 getCellDimensions(scalar3<scalar_t> box_size, scalar_t cutoff) {
+    int3 cell_dim = make_int3(box_size.x / cutoff, box_size.y / cutoff, box_size.z / cutoff);
+    // Minimum 3 cells in each dimension
+    cell_dim.x = thrust::max(cell_dim.x, 3);
+    cell_dim.y = thrust::max(cell_dim.y, 3);
+    cell_dim.z = thrust::max(cell_dim.z, 3);
+// In the host, throw if there are more than 1024 cells in any dimension
+#ifndef __CUDA_ARCH__
+    if (cell_dim.x > 1024 || cell_dim.y > 1024 || cell_dim.z > 1024) {
+        throw std::runtime_error("Too many cells in one dimension. Maximum is 1024");
+    }
+#endif
+    return cell_dim;
+}
+
+/*
+ * @brief Get the cell index of a point
+ * @param p The point position
+ * @param box_size The size of the box in each dimension
+ * @param cutoff The cutoff
+ * @return The cell index
+ */
+template <typename scalar_t>
+__device__ int3 getCell(scalar3<scalar_t> p, scalar3<scalar_t> box_size, scalar_t cutoff) {
+    p = rect::apply_pbc<scalar_t>(p, box_size);
+    // Take to the [0, box_size] range and divide by cutoff (which is the cell size)
+    int cx = floorf((p.x + scalar_t(0.5) * box_size.x) / cutoff);
+    int cy = floorf((p.y + scalar_t(0.5) * box_size.y) / cutoff);
+    int cz = floorf((p.z + scalar_t(0.5) * box_size.z) / cutoff);
+    int3 cell_dim = getCellDimensions(box_size, cutoff);
+    // Wrap around. If the position of a particle is exactly box_size, it will be in the last cell,
+    // which results in an illegal access down the line.
+    if (cx == cell_dim.x)
+        cx = 0;
+    if (cy == cell_dim.y)
+        cy = 0;
+    if (cz == cell_dim.z)
+        cz = 0;
```

Comment:
I must admit sometimes I suffer from early-optimization-itis. 
> traverseCellList() calls it at least 30 times.

Its only two per thread, right?:
https://github.com/torchmd/torchmd-net/blob/0e50bd65b8416cfbdb9b40539748c68a8e3393f2/torchmdnet/neighbors/neighbors_cuda_cell.cuh#L351-L354
You are making me reevaluate a lot of stuff I take for granted, I appreciate it.
I am working under some assumptions here:
1. Arithmetics are free, memory is expensive
2. Constant memory (like the arguments of a kernel) is fast, but only when all threads access it at the same time
3. Modulus is expensive, ternary operator/small branches tend to be faster

This is why I tend to just recompute stuff. In this case placing cell_dim as a kernel argument seems to have no effect in performance. I suspect by declaring cell_dim as const the compiler is actually placing it into a constant memory cache anyway. By recomputing I let the compiler decide whether to place the variable in constant memory or register memory.
Would not be surprised if nvcc is smart enough to see that the same function is called with the same arguments and just skip it.

I did some benchmarks for the modulus thing:
```c++
template <typename scalar_t>
__device__ int3 getCell(scalar3<scalar_t> p, scalar3<scalar_t> box_size, scalar_t cutoff) {
    p = rect::apply_pbc<scalar_t>(p, box_size);
    // Take to the [0, box_size] range and divide by cutoff (which is the cell size)
    int cx = floorf((p.x + scalar_t(0.5) * box_size.x) / cutoff);
    int cy = floorf((p.y + scalar_t(0.5) * box_size.y) / cutoff);
    int cz = floorf((p.z + scalar_t(0.5) * box_size.z) / cutoff);
    int3 cell_dim = getCellDimensions(box_size, cutoff);
    // Wrap around. If the position of a particle is exactly box_size, it will be in the last cell,
    // which results in an illegal access down the line.
    cx = cx == cell_dim.x ? 0 : cx;
    cy = cy == cell_dim.y ? 0 : cy;
    cz = cz == cell_dim.z ? 0 : cz;
    return make_int3(cx, cy, cz);
}
```
With respect to this:
```c++
template <typename scalar_t>
__device__ int3 getCell(scalar3<scalar_t> p, scalar3<scalar_t> box_size, scalar_t cutoff) {
  int3 cell_dim = getCellDimensions(box_size, cutoff);
    int cx = fmodf(floorf((p.x + scalar_t(0.5) * box_size.x) / cutoff), cell_dim.x);
    int cy = fmodf(floorf((p.y + scalar_t(0.5) * box_size.y) / cutoff), cell_dim.y);
    int cz = fmodf(floorf((p.z + scalar_t(0.5) * box_size.z) / cutoff), cell_dim.z);
    return make_int3(cx, cy, cz);
}
```
Or a similar alternative using integer modulus (%). 
```shell
Benchmarking neighbor list generation for 32767 particles with 64 neighbors on average
Batch:                1     2     4     8    16    32    64   128   256   512
Time(ms) modulus: 1.065 1.105 1.290 1.321 1.451 1.869 3.113 3.227 3.526 3.141
Time(ms) ternary: 1.179 1.165 1.366 1.497 1.532 1.863 2.896 3.077 4.151 2.701
Benchmarking neighbor list generation for 1 batches with 64 neighbors on average
N:                  256   512  1024  2048  4096  8192 16384 32768 65536 131072
Time(ms) modulus: 0.134 0.193 0.199 0.197 0.251 0.387 0.583 1.067 2.019 3.713
Time(ms) ternary: 0.141 0.187 0.195 0.212 0.264 0.396 0.605 1.178 2.061 3.748
```

I do not really trust these microbenchmarks because it is bound to be dependent on GPU arch and change with time. However your solution seems faster most time on a 4090 and is much simpler, so lets keep it!
I wonder why large batch sizes make it worse, though.

Reply:
> Arithmetics are free, memory is expensive

It depends on whether your kernel is bound by compute or memory.  If your kernel has efficient memory access patterns and enough occupancy to cover them, compute can easily be the bottleneck.